### PR TITLE
rewrite executor.

### PR DESCRIPTION
**Pull request by hanfei19910905**
Replace "result field" with schema. Evaluate expression with indices of rows during execution.

### CODE REVIEW

**Comment by shenli** [comment]
Please add a few description for this PR. @hanfei19910905

File: optimizer/plan/newplanbuilder.go [hunk 1]
```diff
@@ -405,19 +431,23 @@ func (b *planBuilder) extractAggrFunc(sel *ast.SelectStmt) ([]*ast.AggregateFunc
 	aggrList := extractor.AggFuncs
 	aggrList = append(aggrList, havingAggrFuncs...)
 	aggrList = append(aggrList, orderByAggrFuncs...)
-	totalAggrMapper := make(map[*ast.AggregateFuncExpr]int)
+	totalAggrMapper := make(map[ast.ExprNode]int)
 
 	for i, aggr := range aggrList {
-		totalAggrMapper[aggr] = i
+		if aggr.IsCol {
```
Comment (coocood) [inline optimizer/plan/newplanbuilder.go]: Why use `aggr.Args[0]` as aggregate function?

File: executor/builder.go [hunk 2]
```diff
@@ -212,6 +151,7 @@ func (b *executorBuilder) buildTableScan(v *plan.TableScan) Executor {
 	case "information_schema", "performance_schema":
 		memDB = true
 	}
+	log.Warningf(fmt.Sprintf("old table id %d %s, col id %d %s", table.Meta().ID, table.Meta().Name.L, table.Cols()[0].ID, table.Cols()[0].Name.L))
```
Comment (zimulala) [inline executor/builder.go]: Remove `fmt.Sprintf`

File: optimizer/plan/predicate_push_down.go [hunk 3]
```diff
@@ -168,7 +170,10 @@ func PredicatePushDown(p Plan, predicates []expression.Expression) (ret []expres
 		}
 		return
 	//TODO: support aggregation.
+	case *Aggregation, *Simple:
+		return predicates, nil
 	default:
-		return predicates, errors.Errorf("Unkown type %T.", v)
+		log.Warningf(fmt.Sprintf("Unknown Type %T in Predicate Pushdown", v))
```
Comment (zimulala) [inline optimizer/plan/predicate_push_down.go]: `log.Warningf` is enough.

File: expression/expression.go [hunk 4]
```diff
@@ -137,6 +147,23 @@ func (s Schema) FindColumn(astCol *ast.ColumnName) (*Column, error) {
...
 }
 
+// InitIndices sets indices for columns in schema.
+func (s Schema) InitIndices() {
+	for i, c := range s {
+		c.Index = i
+	}
+}
+
+// RetrieveColumn replace column in expression with column in schema.
```
Comment (zimulala) [inline expression/expression.go]: replace -> replaces ?

File: executor/executor_test.go [hunk 5]
```diff
@@ -1369,3 +1373,36 @@ func (s *testSuite) TestSQLMode(c *C) {
...
+	result = tk.MustQuery("select sum(distinct c) from t group by d")
+	result.Check(testkit.Rows("1", "4", "5"))
+	result = tk.MustQuery("select min(c) from t group by d")
+	result.Check(testkit.Rows("1", "1", "1"))
+	result = tk.MustQuery("select max(c) from t group by d")
+	result.Check(testkit.Rows("1", "3", "4"))
+	result = tk.MustQuery("select avg(c) from t group by d")
+	result.Check(testkit.Rows("1.0000", "2.0000", "2.5000"))
+	plan.UseNewPlanner = false
+
```
Comment (zimulala) [inline executor/executor_test.go]: Remove empty line.

File: optimizer/plan/column_pruning.go [hunk 6]
```diff
@@ -49,36 +113,147 @@ func PruneColumnsAndResolveIndices(p Plan) error {
...
+			} else {
+				rightCols = append(rightCols, col)
+			}
 		}
-	}
-	for _, child := range p.GetChildren() {
-		err := PruneColumnsAndResolveIndices(child)
+		var usedLeft, usedRight []bool
+		var err error
+		usedLeft, err = PruneColumnsAndResolveIndices(p.GetChildByIndex(0), leftCols)
```
Comment (zimulala) [inline optimizer/plan/column_pruning.go]: `=` -> `:=`,   the same as line 222, then remove line 210 and 211.

**Comment by ngaut** [comment]
s/Warningf/Warnf/

File: evaluator/builtin_other.go [hunk 7]
```diff
@@ -0,0 +1,327 @@
+package evaluator
```
Comment (shenli) [inline evaluator/builtin_other.go]: Miss license

File: executor/executor_simple.go [hunk 8]
```diff
@@ -49,6 +50,12 @@ func (e *SimpleExec) Fields() []*ast.ResultField {
 	return nil
 }
 
+// Schema implements Executor Schema interface.
+func (e *SimpleExec) Schema() expression.Schema {
+	return nil
+
```
Comment (shenli) [inline executor/executor_simple.go]: Remove this empty line.

File: executor/new_executor.go [hunk 9]
```diff
@@ -233,7 +239,440 @@ func (e *HashJoinExec) Next() (*Row, error) {
...
+
+// SelectionExec represents a filter executor.
+type SelectionExec struct {
+	Src       Executor
+	Condition expression.Expression
+	ctx       context.Context
+	schema    expression.Schema
+}
+
+// Schema implement Executor Schema interface.
```
Comment (shenli) [inline executor/new_executor.go]: implements

File: evaluator/builtin_other.go [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+		y, err = rightDatum.ToBool()
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		} else if y == 0 {
+			d.SetInt64(y)
+			return
+		}
+	}
+	if leftDatum.Kind() == types.KindNull || rightDatum.Kind() == types.KindNull {
```
Comment (shenli) [inline evaluator/builtin_other.go]: Can we move this to the beginning?

File: evaluator/builtin_other.go [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+		y, err = rightDatum.ToBool()
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		} else if y == 1 {
+			d.SetInt64(y)
+			return
+		}
+	}
+	if leftDatum.Kind() == types.KindNull || rightDatum.Kind() == types.KindNull {
```
Comment (shenli) [inline evaluator/builtin_other.go]: Can we move this to the beginning?

File: evaluator/builtin_other.go [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+				types.KindString,
+				types.KindMysqlDecimal,
+				types.KindBytes,
+				types.KindMysqlHex,
+				types.KindMysqlBit,
+				types.KindMysqlEnum,
+				types.KindMysqlSet:
+				d = aDatum
+			default:
+				err = ErrInvalidOperation
```
Comment (shenli) [inline evaluator/builtin_other.go]: Add some error info, for example ErrInvalidOperation.Gen("Unsupported type %v for opcode.Plus", aDatum.Kind())

File: evaluator/builtin_other.go [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+			case types.KindMysqlEnum:
+				d.SetFloat64(-aDatum.GetMysqlEnum().ToNumber())
+			case types.KindMysqlSet:
+				d.SetFloat64(-aDatum.GetMysqlSet().ToNumber())
+			default:
+				err = ErrInvalidOperation
+				return
+			}
+		default:
+			err = ErrInvalidOperation
```
Comment (shenli) [inline evaluator/builtin_other.go]: Add error info.

File: executor/executor_test.go [hunk 14]
```diff
@@ -686,12 +687,13 @@ func (s *testSuite) TestSelectOrderBy(c *C) {
 		tk.MustExec(fmt.Sprintf("insert INTO select_order_test VALUES (%d, \"zz\");", i))
 	}
 	tk.MustExec("insert INTO select_order_test VALUES (1501, \"aa\");")
+	tk.MustExec("commit")
```
Comment (shenli) [inline executor/executor_test.go]: Why move commit here?

File: executor/new_builder.go [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+	e := &HashJoinExec{
+		schema:      v.GetSchema(),
+		otherFilter: composeCondition(v.OtherConditions),
+		prepared:    false,
+		ctx:         b.ctx,
+	}
+	var leftHashKey, rightHashKey []*expression.Column
+	for _, eqCond := range v.EqualConditions {
+		eqStr, _ := opcode.Ops[opcode.EQ]
+		if eqCond.FuncName.L == eqStr {
```
Comment (shenli) [inline executor/new_builder.go]: Will there be any condition with other opcode?

File: executor/new_builder.go [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+	if err != nil {
+		b.err = err
+		return nil
+	}
+	table, _ := b.is.TableByID(v.Table.ID)
+	client := txn.GetClient()
+	var memDB bool
+	switch v.DBName.L {
+	case "information_schema", "performance_schema":
+		memDB = true
```
Comment (shenli) [inline executor/new_builder.go]: We can return nil directly.

File: executor/executor_test.go (hunk 14 above)
Comment (hanfei19910905) [inline executor/executor_test.go]: because dirty transaction has been implemented yet.

Comment (shenli) [inline executor/executor_test.go]: Got it!

File: evaluator/builtin_other.go [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+					d.SetInt64(zeroI64)
+				}
+			} else {
+				d.SetNull()
+			}
+			return
+		}
+
+		n, err := a.CompareDatum(b)
+
```
Comment (c4pt0r) [inline evaluator/builtin_other.go]: remove this blank line.

File: executor/new_builder.go (hunk 16 above)
Comment (hanfei19910905) [inline executor/new_builder.go]: we still will judge "client.SupportRequestType(kv.ReqTypeSelect, 0) "later

File: executor/new_builder.go [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+
+// compose CNF items into a balance deep CNF tree, which benefits a lot for pb decoder/encoder.
+func composeCondition(conditions []expression.Expression) expression.Expression {
+	length := len(conditions)
+	if length == 0 {
+		return nil
+	} else if length == 1 {
+		return conditions[0]
+	} else {
+		eqStr, _ := opcode.Ops[opcode.AndAnd]
```
Comment (c4pt0r) [inline executor/new_builder.go]: if you don't care whether the value is exists or not, why you use the placeholder?

File: executor/new_builder.go [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+	} else {
+		e.smallExec = b.build(v.GetChildByIndex(1))
+		e.bigExec = b.build(v.GetChildByIndex(0))
+	}
+	return e
+}
+
+func (b *executorBuilder) buildAggregation(v *plan.Aggregation) Executor {
+	src := b.build(v.GetChildByIndex(0))
+	e := &AggregationExec{
```
Comment (c4pt0r) [inline executor/new_builder.go]: return &...{...}

File: executor/new_executor.go [hunk 20]
```diff
@@ -134,25 +151,20 @@ func (e *HashJoinExec) constructMatchedRows(bigRow *Row) (matchedRows []*Row, er
...
 		if e.otherFilter != nil {
-			startKey := 0
-			if !e.leftSmall {
-				startKey = len(bigRow.Data)
-			}
-			for i, data := range smallRow.Data {
-				e.fields[i+startKey].Expr.SetValue(data.GetValue())
-			}
-			otherMatched, err = expression.EvalBool(e.otherFilter, bigRow.Data, e.ctx)
+			otherMatched, err = expression.EvalBool(e.otherFilter, matchedRow.Data, e.ctx)
```
Comment (c4pt0r) [inline executor/new_executor.go]: check the error.

File: executor/new_executor.go [hunk 21]
```diff
@@ -173,19 +185,13 @@ func (e *HashJoinExec) fillNullRow(bigRow *Row) (returnRow *Row, err error) {
 	} else {
 		returnRow = joinTwoRow(bigRow, smallRow)
 	}
-	for i, data := range returnRow.Data {
-		e.fields[i].Expr.SetValue(data.GetValue())
-	}
 	return returnRow, nil
```
Comment (c4pt0r) [inline executor/new_executor.go]: fillNullRow doesn't return a error value.

File: expression/aggregation.go [hunk 22]
```diff
@@ -48,35 +68,261 @@ type aggrCtxMapper map[string]*ast.AggEvaluateContext
...
+	}
+	ctx.Count++
+	return nil
+}
+
+// GetGroupResult implements AggregationFunction interface.
+func (cf *countFunction) GetGroupResult(groupKey []byte) types.Datum {
+	d := types.Datum{}
+	d.SetInt64(cf.getContext(groupKey).Count)
+	return d
```
Comment (c4pt0r) [inline expression/aggregation.go]: it's not the same style with the codes above.

File: evaluator/builtin_other.go [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+func builtinAndAnd(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+	leftDatum := args[0]
+	rightDatum := args[1]
+	if leftDatum.Kind() != types.KindNull {
+		var x int64
+		x, err = leftDatum.ToBool()
+		if err != nil {
+			err = errors.Trace(err)
+			return
```
Comment (zimulala) [inline evaluator/builtin_other.go]: `return d, errors.Trace(err)` is better.

File: evaluator/builtin_other.go [hunk 24]
```diff
@@ -0,0 +1,340 @@
...
+			d.SetInt64(x)
+			return
+		}
+	}
+	if rightDatum.Kind() != types.KindNull {
+		var y int64
+		y, err = rightDatum.ToBool()
+		if err != nil {
+			err = errors.Trace(err)
+			return
```
Comment (zimulala) [inline evaluator/builtin_other.go]: The same as here.

File: evaluator/builtin_other.go [hunk 25]
```diff
@@ -0,0 +1,340 @@
...
+		} else if y == 0 {
+			d.SetInt64(y)
+			return
+		}
+	}
+	if leftDatum.Kind() == types.KindNull || rightDatum.Kind() == types.KindNull {
+		d.SetNull()
```
Comment (zimulala) [inline evaluator/builtin_other.go]: Remove this line.

**Comment by coocood** [comment]
LGTM

File: evaluator/builtin_other.go [hunk 26]
```diff
@@ -0,0 +1,316 @@
...
+			d.SetInt64(zeroI64)
+		}
+		return
+	}
+}
+
+func bitOpFactory(op opcode.Op) BuiltinFunc {
+	return func(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+		a, b := types.CoerceDatum(args[0], args[1])
+
```
Comment (zimulala) [inline evaluator/builtin_other.go]: Remove this blank line.

File: optimizer/plan/newplanbuilder.go [hunk 27]
```diff
@@ -487,46 +510,46 @@ func (b *planBuilder) buildNewSelect(sel *ast.SelectStmt, asName model.CIStr) Pl
 			return nil
 		}
 	}
+	log.Warningf("oldLen %d sel %d", oldLen, len(sel.Fields.Fields))
```
Comment (zimulala) [inline optimizer/plan/newplanbuilder.go]: `Warningf` -> `Warnf`

File: executor/new_builder.go [hunk 28]
```diff
@@ -0,0 +1,157 @@
...
+	client := txn.GetClient()
+	var memDB bool
+	switch v.DBName.L {
+	case "information_schema", "performance_schema":
+		memDB = true
+	}
+	supportDesc := client.SupportRequestType(kv.ReqTypeSelect, kv.ReqSubTypeDesc)
+	if !memDB && client.SupportRequestType(kv.ReqTypeSelect, 0) {
+		// TODO: support condition pushdown and union scan exec.
+		e := &NewTableScanExec{
```
Comment (zimulala) [inline executor/new_builder.go]: return &...{...} directly.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**